File: backend/app/schemas/carousel.py
from typing import Optional
from pydantic import BaseModel

# Shared properties
class CarouselItemBase(BaseModel):
    title: Optional[str] = None
    subtitle: Optional[str] = None
    image: Optional[str] = None
    order: Optional[int] = 0
    button_text: Optional[str] = None
    button_link: Optional[str] = None
    transition_effect: Optional[str] = "slide"
    overlay_effect: Optional[str] = "grid"

# Properties to receive on item creation
class CarouselItemCreate(CarouselItemBase):
    title: str
    image: str
    button_text: Optional[str] = None
    button_link: Optional[str] = None

# Properties to receive on item update
class CarouselItemUpdate(CarouselItemBase):
    pass

# Properties shared by models stored in DB
class CarouselItemInDBBase(CarouselItemBase):
    id: int

    class Config:
        from_attributes = True

# Properties to return to client
class CarouselItem(CarouselItemInDBBase):
    pass

# Properties stored in DB
class CarouselItemInDB(CarouselItemInDBBase):
    pass

File: frontend/src/api/carousel.ts
import api from './axios';

export interface CarouselItem {
    id: number;
    title: string;
    subtitle?: string;
    image: string;
    order: number;
    button_text?: string;
    button_link?: string;
    transition_effect?: string;
    overlay_effect?: string;
}

export interface CarouselItemCreate {
    title: string;
    subtitle?: string;
    image: string;
    order?: number;
    button_text?: string;
    button_link?: string;
    transition_effect?: string;
    overlay_effect?: string;
}

export interface CarouselItemUpdate {
    title?: string;
    subtitle?: string;
    image?: string;
    order?: number;
    button_text?: string;
    button_link?: string;
    transition_effect?: string;
    overlay_effect?: string;
}

export const getCarouselItems = async (): Promise<CarouselItem[]> => {
    const response = await api.get('/carousel/');
    return response.data;
};

export const getCarouselItem = async (id: number): Promise<CarouselItem> => {
    const response = await api.get(`/carousel/${id}`);
    return response.data;
};

export const createCarouselItem = async (data: CarouselItemCreate): Promise<CarouselItem> => {
    const response = await api.post('/carousel/', data);
    return response.data;
};

export const updateCarouselItem = async (id: number, data: CarouselItemUpdate): Promise<CarouselItem> => {
    const response = await api.put(`/carousel/${id}`, data);
    return response.data;
};

export const deleteCarouselItem = async (id: number): Promise<CarouselItem> => {
    const response = await api.delete(`/carousel/${id}`);
    return response.data;
};
